roachtest: add admission-control/disk-bandwidth

irfansharif opened the pull request:
Integration test for disk bandwidth tokens. First commit is from #109350.

Release note: None

irfansharif committed:
server,admission: automatically infer disk names for stores

Part of #86857.

This commit eliminate the need to provide the disk-name common
environments e.g. linux with store on EBS or GCP PD. To make use of AC's
disk bandwidth tokens, users still need to specify the provisioned
bandwidth, for now. So in a sense this machinery is still "disabled by
default". They can also do this through the
kvadmission.store.provisioned_bandwidth cluster setting.

Next steps:

- add roachtests that make use of these disk bandwidth tokens;
- automatically measure provisioned bandwidth, using something like
  github.com/irfansharif/probe, gate behind envvars or cluster settings;
- roll it out in managed environments;
- roll it out elsewhere.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109945)
<!-- Reviewable:end -->

irfansharif committed:
roachtest: add admission-control/disk-bandwidth

Integration test for disk bandwidth tokens, copying over what we ran in
\#82813. Part of #86857

Release note: None

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109945)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @renatolabs, and @smg260)

___
*[`pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109945#-Ndaeyip5oKw3KabjwhM:-Ndaeyip5oKw3KabjwhN:bgfh0zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8be57fcf0d159f765515e489345d816b57804e/pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go#L29)):*
> ```Go
> 
> // This test sets up a 1 node CRDB cluster on an 32vCPU machine, runs low
> // priority kv0 that will overload the compaction throughput (out of L0) of
> ```

We don't want it to saturate compaction throughtput since we now we also have compaction tokens.
To ensure that compaction tokens are not the bottleneck, we should set admission.l0_sub_level_count_overload_threshold, admission.l0_file_count_overload_threshold to very high values, and also increase the compaction concurrency.

Then only the disk bandwidth will be limiting what the elastic work can achieve. But since disk bandwidth is an actual bottleneck, we still need a way to measure that the elastic work got limited by disk bandwidth tokens and not simply by hitting the resource limit. Measuring the regular work throughput may be one way to do this, but since that regular work is closed loop it will some latency change, so this may be a flaky signal. Another possibility would be to see if the WAL sync latency metric shows a clear difference with disk bandwidth tokens disabled and enabled. If that doesn't either we could explicitly reduce the disk bandwidth limit to half the provisioned value and see if that effects the elastic work throughput.
___
*[`pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109945#-Ndaeu9kAVugD29PxwL9:-Ndaeu9kAVugD29PxwLA:bld1eoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8be57fcf0d159f765515e489345d816b57804e/pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go#L31)):*
> ```Go
> // priority kv0 that will overload the compaction throughput (out of L0) of
> // the store. With admission control subjecting this low priority load to
> // elastic IO tokens, the overload is limited.
> ```

elastic disk bandwidth tokens?
___
*[`pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109945#-NdahDX-9Y9p8ayTBUgY:-NdahDX-9Y9p8ayTBUgZ:b-q5bifk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8be57fcf0d159f765515e489345d816b57804e/pkg/cmd/roachtest/tests/admission_control_disk_bandwidth.go#L78)):*
> ```Go
> 
> 				if _, err := db.ExecContext(
> 					ctx, "SET CLUSTER SETTING kvadmission.store.provisioned_bandwidth = '250MiB'"); err != nil {
> ```

how do we know this is the same as the provisioned bandwidth?
If it is higher than the provisioned bandwidth, the bandwidth tokens will not do anything?


<!-- Sent from Reviewable.io -->

irfansharif commented:
@sumeerbhola, @aadityasondhi: Same as https://github.com/cockroachdb/cockroach/pull/109350#issuecomment-1945031709; this is salvageable me-thinks. But I'll close it for now.
